File: leetcode3/황은지/834. Sum of Distances in Tree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var sumOfDistancesInTree = function (n, edges) {
  const dp = Array.from({ length: 6 }, () => Array(6));
  // 미리 2차원 배열에 채운다
  // root에서 내려가면서 한번만 계산하는 방법이 없을까?
  // dfs n번 돌리니 시간초과됨.. dp를 써야될거같은데 방법을 못찾음..
};

File: leetcode3/최원준/834. Sum of Distances in Tree.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(n)]

        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)

        count = [1] * n
        answer = [0] * n

        # 1차 DFS
        # subtree 크기 + answer[0] 계산
        def dfs(node, parent, depth):
            answer[0] += depth

            for child in graph[node]:
                if child == parent:
                    continue

                dfs(child, node, depth + 1)
                count[node] += count[child]

        dfs(0, -1, 0)

        # 2차 DFS
        # root를 parent -> child로 옮기면서 answer 계산
        def reroot(node, parent):
            for child in graph[node]:
                if child == parent:
                    continue

                answer[child] = (
                    answer[node]
                    - count[child]
                    + (n - count[child])
                )

                reroot(child, node)

        reroot(0, -1)

        return answer
